deborphan/doc/deborphan.1
.TH DEBORPHAN 1 "February 2009" "deborphan"
.\" Copyright (C) 2000, 2001, 2002, 2003 Cris van Pelt
.\" Copyright (C) 2003, 2004, 2005, 2006 Peter Palfrader
.\" Copyright (C) 2005 Daniel Déchelotte
.\" Copyright (C) 2008 Andrej Tatarenkow
.\" Copyright (C) 2008, 2009 Carsten Hey

.SH NAME
deborphan \- Orphaned package finder
.SH SYNOPSIS
.B deborphan
[\fIOPTION\fR] \&.\|.\|.\& [\fIPACKAGE\fR] \&.\|.\|.
.SH DESCRIPTION
.I deborphan
finds packages that have no packages depending on them. The default operation
is to search within the libs, oldlibs and introspection sections to hunt down
unused libraries.
.PP
If it is invoked with an optional list of packages, only the dependencies on
those packages will be checked. The results are printed to stdout as if
the option \fI\-\-show-deps\fR had been given. Searching for specific packages
will show the package, regardless of its priority. It is possible to specify
\fI\-\fR, to read a list of packages from standard input.

.SH OPTIONS
.TP
\fB\-f, \-\-status\-file=\fIFILE\fR
Use FILE as the status file.
.TP
\fB\-h, \-\-help\fP
Display a short help message and exit.
.TP
\fB\-v, \-\-version\fP
Display version information and exit.

.\" show stuff
.SS "OUTPUT MODIFIERS"
.TP
\fB\-d, \-\-show\-deps\fP
Show a list of all installed packages and name the packages that depend on
them.
.TP
\fB\-P, \-\-show\-priority\fP
Show the priority of the packages found.
.TP
\fB\-s, \-\-show\-section\fP
Show the sections the packages are in.
.TP
\fB\-\-no\-show\-section\fP
Override showing sections when the default is to show them
(see \fB\-\-all-packages\fR).
.TP
\fB\-z, \-\-show\-size\fP
Show the installed size of the packages found.

.\" search stuff
.SS "SEARCH MODIFIERS"
.TP
\fB\-a, \-\-all\-packages\fP
Check all the packages, instead of only those in the libs section.
Best used (if at all used) in combination with \fB\-\-priority\fR.
This option implies \fB\-\-show-section\fR.
.\" , when compiled with ALL_PACKAGES_IMPLY_SECTION defined (default)
.TP
\fB\-e, \-\-exclude=\fILIST\fR
Excludes the packages named in \fILIST\fR (a comma separated list) from
the evaluation as if they didn't exist in the status file.
.TP
\fB\-H, \-\-force\-hold\fP
Ignore "hold" dpkg-flags on packages and thus display these packages. Without
this option packages with the "hold" flag set will not be displayed.
Please refer to
.BR dpkg (1)
for more information about package flags. Due a bug in
aptitude (Debian bug #137771) hold flags created by aptitude are ignored by
deborphan.
.TP
\fB\-\-ignore\-suggests\fP
.PD 0
.TP
\fB\-\-ignore\-recommends\fP
.PD
Do not check if there is a package `suggesting' (\fB\-\-ignore\-suggests\fP)
or `recommending' (\fB\-\-ignore\-recommends\fP) the package. When both options
are used together, deborphan behaves as if the `nice-mode' mentioned below has
been turned off.
.TP
\fB\-n, \-\-nice\-mode\fP
Turn off nice-mode.
Nice-mode checks if there is a package `suggesting' or `recommending'
the package.
If one is found, the package will be marked as in use, or,
when \fB\-\-show\-deps\fR is used,
print out the package suggesting the package
as if it were depending on it.
.TP
\fB\-p, \-\-priority=\fIPRIORITY\fR
Show only those packages with a priority equal to, or greater than
\fIPRIORITY\fR.
\fIPRIORITY\fR may be in the range of 1\(en5, or one of
\fIrequired\fR, \fIimportant\fR, \fIstandard\fR, \fIoptional\fR,
\fIextra\fR.
Default value for \fIPRIORITY\fR is 2 (\fIimportant\fR).
.TP
\fB\-\-find\-config\fP
This option searches for uninstalled packages which still have configuration
files on the system. It implies the \fB\-a\fR option.
.TP
\fB\-\-libdevel\fP
Also search in section "libdevel".

.\" keep file stuff
.SS "KEEP FILE MANAGEMENT"
.TP
\fB\-A, \-\-add\-keep \fIPKG1\fR \&.\|.\|.\& \fIPKGn\fR
Add packages to the list of packages which are never to be reported, regardless
of their state. You may specify '\fB-\fR' to use standard input. Note that
package names are case-sensitive.
.TP
\fB\-k, \-\-keep\-file=\fIFILE\fR
Use \fIFILE\fR to store the list of kept-back packages.
.TP
\fB\-L, \-\-list\-keep\fP
Show the list of packages that are being kept back.
.TP
\fB\-R, \-\-del\-keep \fIPKG1\fR \&.\|.\|.\& \fIPKGn\fR\fP
Remove packages from the list of packages which are never to be reported.
You may specify '\fB-\fR' to use standard input. If there are no dependencies
for this package next time \fIdeborphan\fR is invoked, it will be reported
again.
.TP
\fB\-Z, \-\-zero\-keep\fP
Purge the entire list of packages that are being kept back. The only option
possible in combination with this option is \fB\-A\fR.

.\" debfoster stuff - not compiled in debian
.\" \fB\-\-df\-keep\fP
.\" Use debfoster's keepfile, regardless of the default setting.
.\" Can not be used if deborphan was compiled without support for debfoster.
.\" .TP
.\" \fB\-\-no\-df\-keep\fP
.\" Do not use debfoster's keepfile.
.\" .TP

.SS "GUESSING"

.\" guessing
.TP
\fB\-\-guess\-*\fP
.PD 0
.TP
\fB\-\-no-guess\-*\fP
.PD
.\" See \fBGUESSING\fP below.
.\" .SH GUESSING
.\" .PP
deborphan can try to guess what packages may not be of much use to you by
examining the package's name and/or description.
It will pretend the package is in the \fImain/libs\fR section,
and report it as if it were a library.
This method is in no way perfect or even reliable,
so beware when using this!
It is also possible to tell deborphan e.g., to guess all interpreters
but not Perl by using \fB\-\-guess\-interpreters\fP
\fB\-\-no\-guess\-perl\fP or to guess all but not Mono by using
\fB\-\-guess-all\fP \fB\-\-no\-guess\-mono\fP.
Please note that the \fB\-\-no\-guess\-\fP option must occur after the
\fB\-\-guess\-\fP option it modifies, this makes it possible to declare
more complex things like to guess all, except interpreters but
additionally try to guess perl.

.PP
The following options are to be prefixed by \fI\-\-guess\-\fR or
(except \fBonly\fP) by \fI\-\-no\-guess\-\fR:

.TP
\fBcommon\fP
This option tries to find common packages, i.e.\& packages with names ending in
\fI\-common\fR.

.TP
\fBdata\fP
This option tries to find data packages, i.e.\& packages with names ending in
\fI\-data\fR.

.TP
\fBdebug\fP
This option tries to find debugging libraries, i.e.\& packages with names ending
in \fI\-dbg\fR or \fI\-dbgsym\fR.

.TP
\fBdev\fP
This option tries to find development packages,
i.e.\& packages with names ending in \fI\-dev\fR.
Also see option \fB\-\-libdevel\fP.

.TP
\fBdoc\fP
This option tries to find documentation packages, i.e.\& packages with names
ending in \fI\-doc\fR.

.TP
\fBdummy\fP
This option tries to find dummy packages, i.e.\& packages with
\fIdummy\fR or \fItransitional\fR in their short description.

.TP
\fBkernel\fP
This option tries to find kernel-modules. It tries to match
.\" Don't hyphenate the next line
\%\fI(\-modules|^nvidia-kernel)\-.*[0\(en9]+\e.[0\(en9]+\e.[0\(en9]+\fR.

.TP
\fBinterpreters\fB
Try to find all interpreter modules (i.e.\& imply \fBruby\fR, \fBpike\fR,
\fBpython\fR, \fBperl\fR and \fBmono\fR).

.TP
\fBmono\fP
This option tries to find mono libraries. It tries to match \fI^libmono\fR.

.TP
\fBperl\fP
This option tries to find perl modules. It tries to match \fI^lib.*\-perl$\fR.

.TP
\fBpike\fP
This option tries to find pike modules. It tries to match
\fI^pike[[:digit:]]*\-\fR.

.TP
\fBpython\fP
This option tries to find python modules. It tries to match
\fI^python[[:digit:]]*\-\fR.

.TP
\fBruby\fP
This option tries to find ruby modules. It tries to match
\fI^lib.*\-ruby[[:digit:].]*$\fR.

.TP
\fBsection\fP
This option tries to find libraries that were accidentally placed in the
wrong section. It tries to match \fI^lib\fR, but not if it ends
in one of: \fI\-dbg\fR, \fI\-dbgsym\fR, \fI\-doc\fR, \fI\-perl\fR, or
\fI\-dev\fR.

.TP
\fBall\fP
Try all of the above.

.TP
\fBonly\fP
Ignore the package's section completely, and just go for the name and/or
description. This option must be used in conjunction with one or more of the
\fB\-\-guess\fR options listed above, or deborphan will not display anything.

.SH FILES
.TP
.I /var/lib/dpkg/status
Statuses of available packages. See the section
.B INFORMATION ABOUT PACKAGES
in \fIdpkg\fR's man-page for more information.
.TP
.I /var/lib/deborphan/keep
A newline-separated list of packages to keep. Package names are in no
particular order.
.SH SEE ALSO
.BR dpkg (8),
.BR dselect (8),
.BR orphaner (8),
.BR editkeep (8),
.BR cruft (8),
.BR xargs (1)
.SH BUGS
If you report a bug, please include your \fI/var/lib/dpkg/status\fR file.
That would help in reproducing the bugs.
.SH AUTHORS

deborphan was written by Cris van Pelt <"Cris van Pelt"@tribe.eu.org>, then
maintained by Peter Palfrader <weasel@debian.org> but is nowadays maintained
Carsten Hey <c.hey@web.de>



deborphan/src/deborphan.c
/* deborphan - Find orphaned libraries on a Debian system.
   Copyright (C) 2000, 2001, 2002, 2003 Cris van Pelt
   Copyright (C) 2003, 2004 Peter Palfrader
   Copyright (C) 2008, 2009 Carsten Hey

   Distributed under the terms of the MIT License, see the
   file COPYING provided in this package for details.
*/

/* Header files we should all have. */
#include <errno.h>
#include <getopt.h>
#include <set.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <config.h>
#include <deborphan.h>

/* These files should already be installed on the host machine -
   the GPL does not allow redistribution in this package. */
#ifdef ENABLE_NLS
#include <libintl.h>
#include <locale.h>
#endif

/* Name this program was called with. */
char* program_name;

/* Options given on the commandline */
int options[NUM_OPTIONS];

/* An optional list of packages to search for. */
dep* search_for;

/* A bunch of packages to keep. */
dep* keep;

static int depcmp(const dep* d1, const dep* d2) {
    return strcmp(d1->name, d2->name);
}

int main(int argc, char* argv[]) {
    char *line, *sfile = NULL, *kfile = NULL;
    char *sfile_content, *sfile_content2;
    pkg_info *package, *this;
    int i, argind;
    size_t j;
    int multiarch = 0;
    int print_arch_suffixes;
    size_t exclude_list_cnt = 0;
    size_t exclude_list_max = 0;
    dep* exclude_list = NULL;

    program_name = argv[0];
    memset(options, 0, NUM_OPTIONS * sizeof(int));

    options[PRIORITY] = DEFAULT_PRIORITY;
#ifdef IGNORE_DEBFOSTER
    options[NO_DEBFOSTER] = 1;
#endif

    /*@unused@*/ /* Actually it is used but splint does not recognise this. */
    struct option longopts[] = {{"version", 0, 0, 'v'},
                                {"help", 0, 0, 'h'},
                                {"status-file", 1, 0, 'f'},
                                {"show-deps", 0, 0, 'd'},
                                {"show-deps-pristine", 0, 0, 'd'},
                                {"nice-mode", 0, 0, 'n'},
                                {"ignore-recommends", 0, 0, 61},
                                {"ignore-suggests", 0, 0, 62},
                                {"print-guess-list", 0, 0, 200},
                                {"check-options", 0, 0, 201},
                                {"all-packages-pristine", 0, 0, 202},
                                {"all-packages", 0, 0, 'a'},
                                {"priority", 1, 0, 'p'},
                                {"show-section", 0, 0, 's'},
                                {"no-show-section", 0, 0, 0},
                                {"show-arch", 0, 0, 203},
                                {"no-show-arch", 0, 0, 204},
                                {"show-priority", 0, 0, 'P'},
                                {"show-size", 0, 0, 'z'},
                                {"force-hold", 0, 0, 'H'},
                                {"keep-file", 1, 0, 'k'},
                                {"add-keep", 0, 0, 'A'},
                                {"del-keep", 0, 0, 'R'},
                                {"list-keep", 0, 0, 'L'},
                                {"zero-keep", 0, 0, 'Z'},
                                {"guess-dev", 0, 0, 1},
                                {"no-guess-dev", 0, 0, 31},
                                {"guess-perl", 0, 0, 2},
                                {"no-guess-perl", 0, 0, 32},
                                {"guess-section", 0, 0, 3},
                                {"no-guess-section", 0, 0, 33},
                                {"guess-all", 0, 0, 4},
                                {"no-guess-all", 0, 0, 34},
                                {"guess-debug", 0, 0, 5},
                                {"no-guess-debug", 0, 0, 35},
#ifdef DEBFOSTER_KEEP
                                {"df-keep", 0, 0, 6},
#endif
                                {"guess-only", 0, 0, 7},
#ifdef DEBFOSTER_KEEP
                                {"no-df-keep", 0, 0, 8},
#endif
                                {"guess-pike", 0, 0, 9},
                                {"no-guess-pike", 0, 0, 39},
                                {"guess-python", 0, 0, 10},
                                {"no-guess-python", 0, 0, 40},
                                {"guess-ruby", 0, 0, 11},
                                {"no-guess-ruby", 0, 0, 41},
                                {"guess-interpreters", 0, 0, 12},
                                {"no-guess-interpreters", 0, 0, 42},
                                {"guess-dummy", 0, 0, 13},
                                {"no-guess-dummy", 0, 0, 43},
                                {"guess-common", 0, 0, 14},
                                {"no-guess-common", 0, 0, 44},
                                {"guess-data", 0, 0, 15},
                                {"no-guess-data", 0, 0, 45},
                                {"guess-doc", 0, 0, 16},
                                {"no-guess-doc", 0, 0, 46},
                                {"find-config", 0, 0, 17},
                                {"libdevel", 0, 0, 18},
                                {"guess-mono", 0, 0, 19},
                                {"no-guess-mono", 0, 0, 49},
                                {"guess-kernel", 0, 0, 20},
                                {"no-guess-kernel", 0, 0, 50},
                                {"guess-java", 0, 0, 21},
                                {"no-guess-java", 0, 0, 51},
                                {"exclude", 1, 0, 'e'},
                                {0, 0, 0, 0}};

#ifdef ENABLE_NLS
    setlocale(LC_ALL, "");
    bindtextdomain(PACKAGE, LOCALEDIR);
    textdomain(PACKAGE);
#endif

    while ((i = getopt_long(argc, argv, "p:advhe:nf:sPzHk:ARLZ", longopts,
                            NULL)) != EOF) {
        switch (i) {
            case 'd':
                options[SHOW_DEPS] = 1;
                break;
            case 'h':
                exit_help();
                break;
            case 'v':
                exit_version();
                break;
            case 'f':
                sfile = (char*)malloc((strlen(optarg) + 1) * sizeof(sfile));
                strcpy(sfile, optarg);
                break;
            case 203:
                options[SHOW_ARCH] = ALWAYS;
                break;
            case 204:
                options[SHOW_ARCH] = NEVER;
                break;
            case 'n':
                options[IGNORE_RECOMMENDS] = 1;
                options[IGNORE_SUGGESTS] = 1;
                break;
            case 61:
                options[IGNORE_RECOMMENDS] = 1;
                break;
            case 62:
                options[IGNORE_SUGGESTS] = 1;
                break;
            case 200:
                printf(
                    "common\ndata\ndebug\ndev\ndoc\ndummy\ninterpreters\nkernel"
                    "\nmono"
                    "\nperl\npike\npython\nruby\nsection\n");
                exit(EXIT_SUCCESS);
            case 201:
                options[CHECK_OPTIONS] = 1;
                break;
            case 202:
                /* ALL_PACKAGES_IMPLY_SECTION is defined anyway, so this
                 * fall through is sufficient for now. */
            case 'a':
                options[ALL_PACKAGES] = 1;
#ifdef ALL_PACKAGES_IMPLY_SECTION
                options[SHOW_SECTION]++;
#endif
                break;
            case 'p':
                options[PRIORITY] = string_to_priority(optarg);
                if (!options[PRIORITY]) {
                    if (!sscanf(optarg, "%d", &options[PRIORITY])) {
                        print_usage(stderr);
                        error(EXIT_FAILURE, 0, "invalid priority: %s", optarg);
                    }
                }
                break;
            case 's':
                options[SHOW_SECTION]++;
                break;
            case 0:
                options[SHOW_SECTION]--;
                break;
            case 'P':
                options[SHOW_PRIORITY] = 1;
                break;
            case 'z':
                options[SHOW_SIZE] = 1;
                break;
            case 'H':
                options[FORCE_HOLD] = 1;
                break;
            case 'k':
                kfile = optarg;
                break;
            case 'A':
                options[ADD_KEEP] = 1;
                break;
            case 'R':
                options[DEL_KEEP] = 1;
                break;
            case 'L':
                options[LIST_KEEP] = 1;
                break;
            case 'Z':
                options[ZERO_KEEP] = 1;
                break;
            case 1:
                guess_set(GUESS_DEV);
                break;
            case 31:
                guess_clr(GUESS_DEV);
                break;
            case 2:
                guess_set(GUESS_PERL);
                break;
            case 32:
                guess_clr(GUESS_PERL);
                break;
            case 3:
                guess_set(GUESS_SECTION);
                break;
            case 33:
                guess_clr(GUESS_SECTION);
                break;
            case 4:
                guess_set(GUESS_ALL);
                break;
            case 34:
                guess_clr(GUESS_ALL);
                break;
            case 5:
                guess_set(GUESS_DEBUG);
                break;
            case 35:
                guess_clr(GUESS_DEBUG);
                break;
#ifdef DEBFOSTER_KEEP
            case 6:
                options[NO_DEBFOSTER] = 0;
                break;
#endif
            case 7:
                options[GUESS_ONLY] = 1;
                break;
#ifdef DEBFOSTER_KEEP
            case 8:
                options[NO_DEBFOSTER] = 1;
                break;
#endif
            case 9:
                guess_set(GUESS_PIKE);
                break;
            case 39:
                guess_clr(GUESS_PIKE);
                break;
            case 10:
                guess_set(GUESS_PYTHON);
                break;
            case 40:
                guess_clr(GUESS_PYTHON);
                break;
            case 11:
                guess_set(GUESS_RUBY);
                break;
            case 41:
                guess_clr(GUESS_RUBY);
                break;
            case 12:
                guess_set(GUESS_IP);
                break;
            case 42:
                guess_clr(GUESS_IP);
                break;
            case 13:
                guess_set(GUESS_DUMMY);
                break;
            case 43:
                guess_clr(GUESS_DUMMY);
                break;
            case 14:
                guess_set(GUESS_COMMON);
                break;
            case 44:
                guess_clr(GUESS_COMMON);
                break;
            case 15:
                guess_set(GUESS_DATA);
                break;
            case 45:
                guess_clr(GUESS_DATA);
                break;
            case 16:
                guess_set(GUESS_DOC);
                break;
            case 46:
                guess_clr(GUESS_DOC);
                break;
            case 17:
                options[FIND_CONFIG] = 1;
                options[ALL_PACKAGES] = 1;
                break;
            case 18:
                options[SEARCH_LIBDEVEL] = 1;
                break;
            case 19:
                guess_set(GUESS_MONO);
                break;
            case 49:
                guess_clr(GUESS_MONO);
                break;
            case 20:
                guess_set(GUESS_KERNEL);
                break;
            case 50:
                guess_clr(GUESS_KERNEL);
                break;
            case 21:
                guess_set(GUESS_JAVA);
                break;
            case 51:
                guess_clr(GUESS_JAVA);
                break;
            case 'e':
                while (optarg) {
                    char* c_ptr;
                    if (exclude_list_cnt >= exclude_list_max) {
                        /* grow exclude_list[] array */
                        exclude_list_max = exclude_list_max
                                               ? exclude_list_max * 2
                                               : INIT_EXCLUDES_COUNT;
#ifdef DEBUG
                        fprintf(stderr, "Growing excludes field to %d.\n",
                                exclude_list_max);
                        fflush(stderr);
#endif /* DEBUG */
                        exclude_list =
                            realloc(exclude_list,
                                    exclude_list_max * sizeof(exclude_list[0]));
                    }
                    if (exclude_list_cnt >= exclude_list_max)
                        error(EXIT_FAILURE, 0, "Too many exclude packages");
                    c_ptr = strsep(&optarg, ",");
                    c_ptr[strcspn(c_ptr, ":")] =
                        '\0'; /* remove architecture suffix */
                    exclude_list[exclude_list_cnt].name = c_ptr;
                    exclude_list[exclude_list_cnt].namehash = strhash(c_ptr);
                    ++exclude_list_cnt;
                }
                qsort(exclude_list, exclude_list_cnt, sizeof(exclude_list[0]),
                      (int (*)(const void*, const void*))depcmp);
                break;
            case '?':
                if (options[CHECK_OPTIONS])
                    exit(EXIT_FAILURE);
                print_usage(stderr);
                exit(EXIT_FAILURE);
        }
    }

    if (options[CHECK_OPTIONS])
        exit(EXIT_SUCCESS);

    if (options[ZERO_KEEP]) {
        if (!kfile)
            kfile = KEEPER_FILE;
        if (zerofile(kfile) < 0)
            error(EXIT_FAILURE, errno, "%s", kfile);

        /* Don't always exit. "deborphan -Z -A <foo>" should be valid. */
        if (!options[ADD_KEEP])
            exit(EXIT_SUCCESS);
    }

    if (options[LIST_KEEP]) {
        if (!kfile)
            kfile = KEEPER_FILE;

        if (!listkeep(kfile)) {
#ifndef DEBFOSTER_KEEP
            if (errno != ENOENT) /* If the file doesn't exit, we just have an
                                    empty list */
                error(EXIT_FAILURE, errno, "%s", kfile); /* fatal */
#else
            fprintf(stderr, "%s: %s: %s\n", program_name, kfile,
                    strerror(errno)); /* non-fatal */
#endif
        }
#ifdef DEBFOSTER_KEEP
        if (!options[NO_DEBFOSTER]) {
            fprintf(stderr, "-- %s -- \n", DEBFOSTER_KEEP);
            listkeep(DEBFOSTER_KEEP);
        }
#endif
        exit(EXIT_SUCCESS);
    }

    if (options[GUESS_ONLY] && !options[GUESS]) {
        print_usage(stderr);
        error(EXIT_FAILURE, 0, "need at least one other --guess option.");
    }

    argind = optind;
    i = 0;
    if (kfile == NULL)
        kfile = KEEPER_FILE;
    if (sfile == NULL)
        sfile = STATUS_FILE;

    if ((argc - argind) > 50)
        error(EXIT_FAILURE, E2BIG, "");

    if (argind < argc) {
        options[SEARCH] = 1;
        options[ALL_PACKAGES] = 1;
        options[SHOW_DEPS] = 1;
        options[PRIORITY] = 0;
    }

    if (options[SHOW_DEPS])
        options[FORCE_HOLD] = 1;

    keep = readkeep(kfile);

    if (options[ADD_KEEP] || options[DEL_KEEP]) {
        char** args;
        if (argind >= argc)
            error(EXIT_FAILURE, 0, "not enough arguments for %s.",
                  options[ADD_KEEP] ? "--add-keep" : "--del-keep");

        args = parseargs(argind, argc, argv);

        for (j = 0; args[j] != NULL; j++)
            args[j][strcspn(args[j], ":")] = '\0'; /* remove arch suffix */

        if (options[DEL_KEEP]) {
            switch (delkeep(kfile, args)) {
                case -1:
                    error(EXIT_FAILURE, errno, "%s", kfile);
                    break; /* make splint happy */
                case 1:
                    error(EXIT_FAILURE, 0, "no packages removed.");
            }
        } else {
            i = pkggrep(sfile, args);
            if (i < 0)
                error(EXIT_FAILURE, errno, sfile);
            if (i)
                error(EXIT_FAILURE, 0, "%s: no such package.", args[i - 1]);

            if ((i = hasduplicate(args)))
                error(EXIT_FAILURE, 0, "%s: duplicate entry.", args[i - 1]);

            if (addkeep(kfile, args) < 0)
                error(EXIT_FAILURE, errno, "%s", kfile);
        }

        exit(EXIT_SUCCESS);
    }

    /* We don't want to merge the files if we're adding, because it's perfectly
       alright to have the same entry in debfoster and deborphan.
    */
#ifdef DEBFOSTER_KEEP
    if (!options[NO_DEBFOSTER]) {
        dep* dfkeep = readkeep(DEBFOSTER_KEEP);

        if (dfkeep) {
            dep* d = keep;
            keep = mergekeep(keep, dfkeep);
            free(d);
            free(dfkeep);
        }
    }
#endif

    search_for = parseargs_as_dep(argind, argc, argv);

    if (!(sfile_content = debopen(sfile)))
        error(EXIT_FAILURE, errno, "%s", sfile);

    sfile_content2 = sfile_content;

    this = package = (pkg_info*)malloc(sizeof(pkg_info));
    init_pkg(this);
    init_pkg_regex();

    while ((line = strsep(&sfile_content, "\n")) != NULL) {
        if ((!strchr("AIPpSsEeDdRr", *line)) || (*line == ' '))
            continue;

        if (*line != '\0') {
            strstripchr(line, ' ');
            get_pkg_info(line, this, &multiarch);
        } else {
            if ((this->self.name && exclude_list &&
                 bsearch(&this->self, exclude_list, exclude_list_cnt,
                         sizeof(exclude_list[0]),
                         (int (*)(const void*, const void*))depcmp)) ||
                (!this->install && !options[FIND_CONFIG])) {
                reinit_pkg(this);
            } else {
                this->next = malloc(sizeof(pkg_info));
                this = this->next;
                init_pkg(this);
            }
        }
    }

    free(sfile_content2);

    this->next = NULL;
    this = package;

    print_arch_suffixes = (options[SHOW_ARCH] == ALWAYS ||
                           (options[SHOW_ARCH] == DEFAULT && multiarch));

    /* Check the dependencies. Last package is handled specially, because it
       is not terminated with a \n in STATUS_FILE. */
    while (this->next) {
        check_lib_deps(package, this, print_arch_suffixes);
        this = this->next;
    }
    if (this->install)
        check_lib_deps(package, this, print_arch_suffixes);

    free_pkg_regex();

    fflush(stdout);

    for (i = 0; options[SEARCH] && search_for[i].name; i++) {
        fprintf(stderr, "%s: package %s", argv[0], search_for[i].name);
        if (search_for[i].arch)
            fprintf(stderr, ":%s", search_for[i].arch);
        fprintf(stderr, " not found or not installed\n");
    }

    return i ? EXIT_FAILURE : EXIT_SUCCESS;
}
